xgboost_ray/data_sources/dask.py

```diff
@@ -0,0 +1,153 @@
+
+        # Dask does not support iloc() for row selection, so we have to
+        # compute a local pandas dataframe first
+        local_df = data.compute()
+
+        if indices:
+            local_df = local_df.iloc[indices]
+
+        if ignore:
+            local_df = local_df[local_df.columns.difference(ignore)]
+
+        return local_df
+
+    @staticmethod
+    def convert_to_series(data: Any) -> pd.Series:
+        _assert_dask_installed()
+        from dask.dataframe import DataFrame as DaskDataFrame, \
+            Series as DaskSeries
+        from dask.array import Array as DaskArray
+
+        if isinstance(data, DaskDataFrame):
+            return pd.Series(data.compute().squeeze())
+        elif isinstance(data, DaskSeries):
+            return data.compute()
+        elif isinstance(data, DaskArray):
+            return pd.Series(data.compute())
+
+        return DataSource.convert_to_series(data)
+
+    @staticmethod
+    def get_actor_shards(
+            data: Any,  # dask.dataframe.DataFrame
+            actors: Sequence[ActorHandle]) -> \
+            Tuple[Any, Optional[Dict[int, Any]]]:
+        _assert_dask_installed()
+
+        actor_rank_ips = get_actor_rank_ips(actors)
+
+        # Get IPs and partitions
+        ip_to_parts = get_ip_to_parts(data)
+
+        return data, assign_partitions_to_actors(ip_to_parts, actor_rank_ips)
+
+    @staticmethod
+    def get_n(data: Any):
+        """
+        For naive distributed loading we just return the number of rows
+        here. Loading by shard is achieved via `get_actor_shards()`
+        """
+        return len(data)
+
+
+def get_ip_to_parts(data: Any) -> Dict[int, Sequence[Any]]:
+    persisted = data.persist(scheduler=ray_dask_get)
+    name = persisted._name
+
+    node_ids_to_node = {node["NodeID"]: node for node in ray.state.nodes()}
+
+    # This is a hacky way to get the partition node IDs, and it's not
+    # 100% accurate as the map task could get scheduled on a different node
+    # (though Ray tries to keep locality). We need to use that until
+    # ray.state.objects() or something like it is available again.
+    partition_locations_df = persisted.map_partitions(lambda df: pd.DataFrame(
+        [ray.get_runtime_context().node_id.hex()])).compute()
+    partition_locations = [
+        partition_locations_df[0].iloc[i]
+        for i in range(partition_locations_df.size)
+    ]
+
+    ip_to_parts = defaultdict(list)
+    for (obj_name,
+         pid), obj_ref in dask.base.collections_to_dsk([persisted]).items():
+        assert obj_name == name
+
+        if isinstance(obj_ref, ray.ObjectRef):
+            node_id = partition_locations[pid]
+            node = node_ids_to_node.get(node_id, {})
+            ip = node.get("NodeManagerAddress", "_no_ip")
+        else:
+            ip = "_no_ip"
+
+        # Pass tuples here (integers can be misinterpreted as row numbers)
+        ip_to_parts[ip].append((pid, ))
```

Comment:
hmm, why would it be misinterpreted?

Reply:
In `load_data()`, we are doing this check:

```

        if indices is not None and len(indices) > 0 and isinstance(
                indices[0], Tuple):
```

Usually, numerical indices would indicate the row numbers a worker should load. This is also valid for Dask (e.g. for centralized loading). However, in this case the numbers should indicate the partition numbers. To not confuse these with row indices, we use a different type. IMO tuples are a good choice because they are immutable and should produce very few overhead.

Note that in Modin we use Ray ObjectIDs instead, but we don't have these for Dask at this point in the code.